File: crate-common/src/v1/types/room_analytics.rs
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "utoipa")]
use utoipa::{IntoParams, ToSchema};

use crate::v1::types::{misc::Time, ChannelId, InviteCode, UserId};

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
pub enum Aggregation {
    Hourly,
    Daily,
    Weekly,
    Monthly,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema, IntoParams))]
pub struct RoomAnalyticsParams {
    pub start: Option<Time>,
    pub end: Option<Time>,
    pub aggregate: Aggregation,

    /// limit between 1..1024, default to 10
    pub limit: Option<u16>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
pub struct RoomAnalyticsMembersCount {
    /// The bucket for this data point.
    pub bucket: Time,

    /// Total number of members in this room.
    pub count: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
pub struct RoomAnalyticsMembersJoin {
    /// The bucket for this data point.
    pub bucket: Time,

    /// Total number of members who joined this room.
    pub count: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
pub struct RoomAnalyticsMembersLeave {
    /// The bucket for this data point.
    pub bucket: Time,

    /// Total number of members who left this room.
    pub count: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema, IntoParams))]
pub struct RoomAnalyticsChannelParams {
    /// return only analytics for this channel, otherwise return data points for everything
    pub channel_id: Option<ChannelId>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
pub struct RoomAnalyticsChannel {
    /// The bucket for this data point.
    pub bucket: Time,
    pub channel_id: ChannelId,
    pub message_count: u64,
    pub media_count: u64,
    pub media_size: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
pub struct RoomAnalyticsOverview {
    /// The bucket for this data point.
    pub bucket: Time,

    /// number of messages sent
    pub message_count: u64,

    /// number of files sent
    pub media_count: u64,

    /// number of files sent
    pub media_size: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
pub struct RoomAnalyticsInvites {
    /// The bucket for this data point.
    pub bucket: Time,

    /// where this member came from
    pub origin: RoomAnalyticsInvitesOrigin,

    /// number of times this invite was used
    pub uses: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "utoipa", derive(ToSchema))]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize), serde(tag = "type"))]
pub enum RoomAnalyticsInvitesOrigin {
    /// user joined with this invite code
    Invite { code: InviteCode },

    /// this was a bot that was installed manually
    BotInstall,

    /// this is a puppet user and was added by a bridge
    Bridged {
        /// the bridge that owns this puppet
        bridge_id: UserId,
    },

    /// user joined directly
    PublicJoin,

    /// unknown or anonymized origin
    Other,
}

File: frontend/src/api/room_analytics.ts
import type { Api } from "@/api";
import type {
	RoomAnalyticsChannel,
	RoomAnalyticsInvites,
	RoomAnalyticsMembersCount,
	RoomAnalyticsMembersJoin,
	RoomAnalyticsMembersLeave,
	RoomAnalyticsOverview,
	Time,
} from "sdk";

export type Aggregation = "Hourly" | "Daily" | "Weekly" | "Monthly";

export type RoomAnalyticsParams = {
	start?: Time;
	end?: Time;
	aggregate: Aggregation;
	limit?: number;
};

export class RoomAnalytics {
	api: Api = null as unknown as Api;

	async getOverview(
		room_id: string,
		query: RoomAnalyticsParams,
	): Promise<RoomAnalyticsOverview[]> {
		const { data, error } = await (this.api.client.http as any).GET(
			"/api/v1/room/{room_id}/analytics/overview",
			{
				params: {
					path: { room_id },
					query: {
						start: query.start,
						end: query.end,
						aggregate: query.aggregate,
						limit: query.limit,
					},
				},
			},
		);
		if (error) throw error;
		return data;
	}

	async getMembersCount(
		room_id: string,
		query: RoomAnalyticsParams,
	): Promise<RoomAnalyticsMembersCount[]> {
		const { data, error } = await (this.api.client.http as any).GET(
			"/api/v1/room/{room_id}/analytics/members-count",
			{
				params: {
					path: { room_id },
					query: {
						start: query.start,
						end: query.end,
						aggregate: query.aggregate,
						limit: query.limit,
					},
				},
			},
		);
		if (error) throw error;
		return data;
	}

	async getMembersJoin(
		room_id: string,
		query: RoomAnalyticsParams,
	): Promise<RoomAnalyticsMembersJoin[]> {
		const { data, error } = await (this.api.client.http as any).GET(
			"/api/v1/room/{room_id}/analytics/members-join",
			{
				params: {
					path: { room_id },
					query: {
						start: query.start,
						end: query.end,
						aggregate: query.aggregate,
						limit: query.limit,
					},
				},
			},
		);
		if (error) throw error;
		return data;
	}

	async getMembersLeave(
		room_id: string,
		query: RoomAnalyticsParams,
	): Promise<RoomAnalyticsMembersLeave[]> {
		const { data, error } = await (this.api.client.http as any).GET(
			"/api/v1/room/{room_id}/analytics/members-leave",
			{
				params: {
					path: { room_id },
					query: {
						start: query.start,
						end: query.end,
						aggregate: query.aggregate,
						limit: query.limit,
					},
				},
			},
		);
		if (error) throw error;
		return data;
	}

	async getChannels(
		room_id: string,
		query: RoomAnalyticsParams,
		channel_id?: string,
	): Promise<RoomAnalyticsChannel[]> {
		const { data, error } = await (this.api.client.http as any).GET(
			"/api/v1/room/{room_id}/analytics/channels",
			{
				params: {
					path: { room_id },
					query: {
						start: query.start,
						end: query.end,
						aggregate: query.aggregate,
						limit: query.limit,
						channel_id,
					},
				},
			},
		);
		if (error) throw error;
		return data;
	}
}
